docs/src/adr/018adr-inlining.md

```diff
@@ -0,0 +1,73 @@
+# ADR-018: Inlining
+
+| author     | revision | last revised |
+| ------------ | --------:| -------- |
+| Jure Kukovec |    1 | Apr 13, 2022 |
+
+**Table of Contents**
+
+- [Summary (Overview)](#summary)
+- [Context (Problem)](#context)
+- [Options (Alternatives)](#options)
+- [Solution (Decision)](#solution)
+- [Consequences (Retrospection)](#consequences)
+
+## Summary
+
+<!-- Statement to summarize, following the following formula: -->
+
+This ADR outlines the transformations that happen in the Apalache inliner pass.
+Since we have recently reworked the inliner in #1569, we saw it fit to document exactly how inlining is supposed to work and why the inlining pass is the way it is. 
+
+## Context
+
+<!-- Communicates the forces at play (technical, political, social, project).
+     This is the story explaining the problem we are looking to resolve.
+-->
+The term "inlining" typically refers to the process of replacing instances of operator application `A(x1,...,xn)`, with `e[x1/p1,...,xn/pn]` - the expression obtained by replacing each instance of `pi` with `xi` -  where the operator `A` is defined as `A(p1,...,pn) == e` and is _non-recursive_.
+We elect to use the term in a broader sense of "replacing an operator with its definition", in the two ways listed below:
+  1. Standard inlining: the instantiation described above
+      1. Non-nullary inlining: the instantiation described above, except the inlining skips nullary LET-defined operators
+  2. Pass-by-name inlining: replacing an operator name `A` with a local LET-definition: `LET A_LOCAL(p1,...,pn) == A(p1,...,pn) IN A_LOCAL`
```

Comment:
Before giving these definitions, please introduce "nullary LET-defined operators" and explain the need for local LET-definitions.

Reply:
> Why is 2. considered inlining at all? Do I understand correctly that this is essentially just eta expansion (i.e., converting `𝜆x.Mx` to `M`) only via a let expression?

It's inlinining in the sense that 
> We elect to use the term in a broader sense of "replacing an operator with its definition"

You can also view it as localization. Suppose you have an expression `e`, in which `A` appears as an argument to a HO operator (e.g. `FoldSet(A, v, S)`). The treatment of this expression depends on the scope surrounding `e`, that is, the definition of `A` is external to `e`. In the `Fold` example, `A` is supposed to be "applied" to the arguments, but we don't know what the arguments are until execution. (2) makes it so the definition is present where the name is passed as an argument, meaning `Fold` can now be analyzed locally, w/o needing to track "scope".

Reply:
So I guess the real kernel of my question is that

> We elect to use the term in a broader sense of "replacing an operator with its definition"

does not seems accurate to the spirit of the phrase "inlining" if we are replacing it with newly synthesized definitions. This seems like a distinct kind of rewriting to me (again, it looks to me like eta expansion). There might not be any stakes to this difference tho, so I'm ok with leaving it as is, even if I personally find the terminology a bit misleading.